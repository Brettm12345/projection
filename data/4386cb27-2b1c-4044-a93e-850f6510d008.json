{
  "user": "hlinner",
  "repo": "doom-emacs",
  "source": "github"
}