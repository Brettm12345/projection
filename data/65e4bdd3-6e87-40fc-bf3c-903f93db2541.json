{"name":"doom-emacs","meta":{"user":"hlinner","repo":"doom-emacs","source":"github"}}