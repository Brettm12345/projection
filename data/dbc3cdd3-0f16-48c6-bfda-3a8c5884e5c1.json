{
  "user": "brettm12345",
  "repo": "xmonad-config",
  "source": "github"
}