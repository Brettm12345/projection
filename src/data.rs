use colored::*;
use fp_core::chain::Chain;
use git2::Repository;
use serde::{Deserialize, Serialize};
use std::fmt::{self, Display};
use std::path::{Path, PathBuf};
use std::str::FromStr;
use url::{ParseError, Url};

pub trait ToUrl {
    fn to_url(&self) -> Result<Url, ParseError>;
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Source {
    BitBucket,
    GitLab,
    GitHub,
}

impl ToString for Source {
    fn to_string(&self) -> String {
        match &self {
            Source::BitBucket => "bb",
            Source::GitLab => "gl",
            Source::GitHub => "gh",
        }
        .to_string()
    }
}

impl FromStr for Source {
    type Err = String;
    fn from_str(s: &str) -> Result<Source, Self::Err> {
        match s {
            "gh" => Ok(Source::GitHub),
            "gl" => Ok(Source::GitLab),
            "bb" => Ok(Source::BitBucket),
            _ => Err(format!(
                "Failed to parse {}, accepted inputs are gh|gl|bb",
                s
            )),
        }
    }
}

impl ToUrl for Source {
    fn to_url(&self) -> Result<Url, ParseError> {
        let l = match self {
            Source::GitHub => "github.com",
            Source::GitLab => "gitlab.com",
            Source::BitBucket => "bitbucket.com",
        };
        Url::parse(&format!("https://{}", l))
    }
}

#[derive(Clone, PartialEq, Debug, Serialize, Deserialize)]
pub struct Project {
    pub user: String,
    pub repo: String,
    pub source: Source,
}

impl FromStr for Project {
    type Err = String;
    fn from_str(s: &str) -> Result<Project, Self::Err> {
        let mut location = s.split(':');
        let source = location.nth(0).chain(|s| Source::from_str(s).ok());
        let mut p = location.nth(0).map(|s| s.split("/")).unwrap();
        let user = p.nth(0).map(String::from);
        let path = p.nth(0).map(String::from);
        Ok(Project {
            user: user.unwrap(),
            source: source.unwrap(),
            repo: path.unwrap(),
        })
    }
}

impl ToUrl for Project {
    fn to_url(&self) -> Result<Url, ParseError> {
        Url::parse(&format!(
            "{}{}/{}",
            self.source.to_url()?.as_str(),
            self.user,
            self.repo
        ))
    }
}

pub trait ToPath {
    fn to_path(&self) -> PathBuf;
}

impl ToPath for Project {
    fn to_path(&self) -> PathBuf {
        Path::new(&format!(
            "{}--{}--{}",
            self.source.to_string(),
            self.user,
            self.repo
        ))
        .to_path_buf()
    }
}

impl Display for Project {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}/{}", &self.user.blue(), &self.repo.purple())
    }
}

pub trait CloneRepo {
    fn clone_repo(&self, root: &str) -> Result<Repository, git2::Error>;
}

impl CloneRepo for Project {
    fn clone_repo(&self, root: &str) -> Result<Repository, git2::Error> {
        Repository::clone(
            &format!("{}", &self.to_url().unwrap().as_str()),
            format!("{}/{}", root, self.to_path().to_str().unwrap()),
        )
    }
}
#[cfg(test)]

mod tests {
    use super::*;
    use insta::assert_debug_snapshot;
    use pretty_assertions::assert_eq;

    #[test]
    fn parse_from_json() {
        let result: Project = serde_json::from_str(
            r#"
        {
            "source": "github",
            "user": "brettm12345",
            "repo": "projection"
        }"#,
        )
        .unwrap();
        assert_debug_snapshot!(result)
    }
    #[test]
    fn convert_to_url() {
        assert_debug_snapshot!(Project {
            user: "brettm12345".to_owned(),
            repo: "projection".to_owned(),
            source: Source::GitHub,
        }
        .to_url())
    }
    #[test]
    fn parse_from_input() {
        assert_eq!(
            Project::from_str("gh:brettm12345/xmonad-config"),
            Ok(Project {
                user: "brettm12345".to_owned(),
                repo: "xmonad-config".to_owned(),
                source: Source::GitHub
            })
        )
    }
}
