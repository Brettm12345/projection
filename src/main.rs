mod cli;
use colored::*;
mod data;
use data::{CloneRepo, Ensure, Project, ToPath};
use dialoguer::Confirmation;
use enquirer::ColoredTheme;
use fuzzy_matcher::skim::SkimMatcherV2;
use fuzzy_matcher::FuzzyMatcher;
use jfs::Store;
use skim::{Skim, SkimOptionsBuilder};
use std::collections::BTreeMap;
use std::fs;
use std::io::Cursor;
use std::iter::{FromIterator, Iterator};
use std::path::Path;
use std::str::FromStr;

type Projects = BTreeMap<String, Project>;

fn select_author(projects: Projects, author: &str) -> Projects {
    let mut res = Projects::new();
    for (id, project) in projects
        .iter()
        .filter(|(_, project)| project.user.eq(author))
    {
        res.insert(id.to_owned(), project.to_owned());
    }
    res
}

fn main() {
    let app = cli::build_cli();
    let matches = app.get_matches();
    let mut cfg = jfs::Config::default();
    cfg.pretty = true;
    let db = Store::new_with_cfg("data", cfg).unwrap();
    let projects: Projects = db
        .all()
        .map(|p| match matches.value_of("author") {
            Some(author) => select_author(p, author),
            None => p,
        })
        .unwrap();

    let project_dir = Path::new(matches.value_of("project-directory").unwrap());

    let fuzzy_search = |q: &str| -> Vec<&Project> {
        let matcher = SkimMatcherV2::default();
        let fuzz = |p: &Project| -> Option<i64> { matcher.fuzzy_match(&format!("{}", p), q) };
        let mut v = Vec::from_iter(projects.values().filter(|p| fuzz(p).is_some()));
        v.sort_by(|&a, &b| fuzz(b).cmp(&fuzz(a)));
        v
    };
    let project_string = projects
        .iter()
        .fold("".to_string(), |line, (_, p)| format!("{}{}\n", line, p));

    let find = |string: &str| -> Option<(&String, &Project)> {
        projects
            .iter()
            .find(|(_, project)| project.repo.as_str().contains(string))
    };

    let confirm = |string: &str| {
        Confirmation::with_theme(&ColoredTheme::default())
            .with_text(string)
            .interact()
            .unwrap_or(false)
    };

    match matches.subcommand() {
        ("add", Some(m)) => {
            let project = Project::from_str(m.value_of("source").unwrap()).unwrap();
            project.clone_repo(project_dir).unwrap();
            match db.save(&project) {
                Ok(_) => println!("{} added {} to projects", "Successfully".green(), project),
                _ => println!("{} adding {} to projects", "Error".red(), project),
            }
        }
        ("remove", Some(m)) => {
            let query = m.value_of("name").unwrap();
            match find(query) {
                Some((id, project)) => {
                    if confirm(&format!(
                        "Are you sure you want to remove {} from your projects?",
                        project
                    )) {
                        if confirm("Also remove the project directory") {
                            match fs::remove_dir_all(project_dir.join(project.to_path())) {
                                Ok(()) => println!("Deleted {}", project),
                                _ => println!("Failed to remove dir project files"),
                            }
                        }
                        match db.delete(id) {
                            Ok(()) => println!("Removed from project list {}", id.cyan()),
                            err => println!(
                                "Failed to remove {}\n{}: {:?}",
                                project,
                                "Error".red(),
                                err
                            ),
                        }
                    }
                }
                None => println!("Failed to find {} in projects", query),
            }
        }
        ("check", _) => {
            for (_, project) in projects.iter() {
                match project.ensure(project_dir) {
                    Ok(_) => continue,
                    _ => {
                        if confirm(&format!(
                            "Project {} is missing. Would you like to clone it",
                            project
                        )) {
                            project.clone_repo(project_dir).unwrap();
                        }
                    }
                }
            }
        }
        ("path", Some(m)) => match find(m.value_of("name").unwrap()) {
            Some((_, project)) => println!(
                "{}",
                project_dir
                    .join(project.to_path())
                    .as_path()
                    .to_str()
                    .unwrap()
            ),

            _ => println!("Unable to find item"),
        },
        ("select", _) => {
            let selected_items = Skim::run_with(
                &SkimOptionsBuilder::default()
                    .height(Some("50%"))
                    .multi(true)
                    .ansi(true)
                    .build()
                    .unwrap(),
                Some(Box::new(Cursor::new(project_string))),
            )
            .map(|out| out.selected_items)
            .unwrap_or_else(Vec::new);

            for item in selected_items.iter() {
                println!("{:#?}", item);
            }
        }
        ("search", Some(m)) => fuzzy_search(m.value_of("query").unwrap_or_else(|| ""))
            .iter()
            .for_each(|project| println!("{}", project)),
        _ => println!("{}", project_string),
    };
}

#[cfg(test)]
mod tests {
    use assert_cmd::Command;
    use insta::assert_debug_snapshot;
    use tempfile::tempdir;

    #[test]
    fn add_project() {
        let project_dir = tempdir().unwrap();
        assert_debug_snapshot!(Command::cargo_bin("projection")
            .unwrap()
            .arg("-d")
            .arg(&format!("{}", project_dir.path().display()))
            .arg("add")
            .arg("gh:brettm12345/xmonad-config")
            .assert());
        assert!(dir_diff::is_different(project_dir.path(), tempdir().unwrap().path()).unwrap())
    }
}
