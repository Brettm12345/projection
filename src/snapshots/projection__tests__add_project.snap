---
source: src/main.rs
expression: "cmd.arg(\"-d\").arg(dir.path().to_owned()).arg(\"add\").arg(\"gh:brettm12345/projection\").assert()"
---
Assert {
    output: Output {
        status: ExitStatus(
            ExitStatus(
                0,
            ),
        ),
        stdout: "Successfully added brettm12345/projection to projects\n",
        stderr: "",
    },
}
